qiime2/plugin/plugin.py

```diff
@@ -201,9 +205,14 @@ def register_semantic_type_to_format(self, semantic_type, artifact_format):
                 raise ValueError("%r has a predicate, differentiating format"
                                  " on predicate is not supported.")
 
-        self.type_formats.append(TypeFormatRecord(
+        record = TypeFormatRecord(
             type_expression=semantic_type,
-            format=artifact_format, plugin=self))
+            format=artifact_format, plugin=self)
+
+        self.type_formats.append(record)
+        for type_ in record.type_expression:
+            self.types.add(SemanticTypeRecord(
```

Comment:
I would match the older `self.types` structure here (using a dict) as it make life marginally easier if you don't care about the `SemanticTypeRecord` objects (e.g. `list(self.types)` vs `[r.semantic_type for r in self.types]`).

This is a good approach otherwise.